ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+     * @throws IOException
+     * @throws MessageFormatException
+     */
+    @Override
+    void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
+        throws IOException, MessageFormatException {
+      // If we find blob metadata in frontend cache, validate it and use it.
+      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      /**
+       * Some cached variables are mutable.
+       * Update them and then compare if other immutable variables are same or not.
+       */
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      blobInfo.setLifeVersion(lifeVersion);
+      lifeVersion = messageInfo.getLifeVersion();
+      if (!serverBlobProperties.equals(receivedBlobProperties)) {
+        /**
+         * This is a rare error case.
+         * If we find blob properties is in our frontend cache, we request just blobInfo from backend to validate
+         * the cache entry. If cached blob properties does not match received blob properties, due to a random bit flip for eg.,
+         * we must delete the cache entry and throw a retriable exception to the user.
+         * The next request will re-fill the cache with a valid entry if the blob stills exists.
+         * We do not need to decrypt anything at this point.
+         */
+        String reason = "Cached blob property does not match received blob property";
+        logger.error("{} for blobId = {}, cached blob property = {}, received blob property = {}", reason, blobId,
+            serverBlobProperties, receivedBlobProperties);
+        deleteMetadata(reason);
+        onInvalidCacheEntry(new RouterException(reason, RouterErrorCode.UnexpectedInternalError));
+        return;
+      }
+      getOptions().ageAtAccessTracker.trackAgeAtAccess(serverBlobProperties.getCreationTimeInMs());
+
+      /**
+       * If we reach here, then the cached metadata is valid and already decrypted from a previous GET/RANGE request.
+       * Just use it directly and initiate fetching data chunks.
+       */
+      requestDataChunks();
+    }
+
+    /**
+     * Resolve range, if needed, and request data chunks
+     * @throws IllegalArgumentException
+     */
+    void requestDataChunks() throws IllegalArgumentException {
+      long dataRangeSize = totalSize;
+      if (options.getBlobOptions.hasBlobSegmentIdx()) {
+        int requestedSegment = options.getBlobOptions.getBlobSegmentIdx();
+        if (requestedSegment < 0 || requestedSegment >= chunkMetadataList.size()) {
+          throw new IllegalArgumentException(
+              "Bad segment number: " + requestedSegment + ", num of keys: " + chunkMetadataList.size());
+        }
+        chunkMetadataList = chunkMetadataList.subList(requestedSegment, requestedSegment + 1);
+        dataRangeSize = chunkMetadataList.get(0).getSize();
+      }
+      if (resolveRangeProper(dataRangeSize)) {
```

Comment:
If we decide to look up cache for segmented GETs too in `shouldLookupMetadataCache`, we might need to check if this a range request by `if (options.getBlobOptions.getRange() != null) ` here.

Reply:
agreed, i need to read up a bit on segmented GETs

Reply:
For this, we only reach this code for a range request. So no need to check getRange() != null. This a policy controlled by shouldLookUpMetadataCache()

Reply:
ok updated code to use cache for segmented GETs also